Source/hydro/Castro_mol_nd.F90

```diff
@@ -59,66 +60,238 @@ subroutine ca_mol_plm_reconstruct(lo, hi, &
+                              HALF*dx(2)*q(i,j,k,QRHO)*const_grav
 
+                         ! right state at i-1/2 interface
+                         qp(i,j,k,n,2) = q(i,j,k,n) - HALF*dq(i,j,k,n) - &
+                              HALF*dx(2)*q(i,j,k,QRHO)*const_grav
+                      end do
                    end do
                 end do
-             end do
+
+             else
+
+                do k = lo(3), hi(3)
+                   do j = lo(2), hi(2)
+                      do i = lo(1), hi(1)
+
+                         ! left state at j+1/2 interface
+                         qm(i,j+1,k,n,2) = q(i,j,k,n) + HALF*dq(i,j,k,n)
+
+                         ! right state at j-1/2 interface
+                         qp(i,j,k,n,2) = q(i,j,k,n) - HALF*dq(i,j,k,n)
+
+                      end do
+                   end do
+                end do
+             end if
 #endif
 
 #if BL_SPACEDIM == 3
           else
 
-             do k = lo(3), hi(3)
-                do j = lo(2), hi(2)
-                   do i = lo(1), hi(1)
+             if (plm_well_balanced == 1 .and. n == QPRES .and. idir == AMREX_SPACEDIM) then
+
+                do k = lo(3), hi(3)
+                   do j = lo(2), hi(2)
+                      do i = lo(1), hi(1)
+
+                         ! left state at i+1/2 interface
+                         qm(i,j,k+1,n,3) = q(i,j,k,n) + HALF*dq(i,j,k,n) + &
+                              HALF*dx(3)*q(i,j,k,QRHO)*const_grav
+
+                         ! right state at i-1/2 interface
+                         qp(i,j,k,n,3) = q(i,j,k,n) - HALF*dq(i,j,k,n) - &
+                              HALF*dx(3)*q(i,j,k,QRHO)*const_grav
+                      end do
+                   end do
+                end do
+
+             else
+
+                do k = lo(3), hi(3)
+                   do j = lo(2), hi(2)
+                      do i = lo(1), hi(1)
 
-                      ! left state at k+1/2 interface
-                      qm(i,j,k+1,n,3) = q(i,j,k,n) + HALF*dq(i,j,k,n)
+                         ! left state at k+1/2 interface
+                         qm(i,j,k+1,n,3) = q(i,j,k,n) + HALF*dq(i,j,k,n)
 
-                      ! right state at k-1/2 interface
-                      qp(i,j,k,n,3) = q(i,j,k,n) - HALF*dq(i,j,k,n)
+                         ! right state at k-1/2 interface
+                         qp(i,j,k,n,3) = q(i,j,k,n) - HALF*dq(i,j,k,n)
 
+                      end do
                    end do
                 end do
+
+             end if
+#endif
+
+          end if
+
+       end do ! component loop
+
+       ! special care for reflecting BCs
+       if (idir == 1) then
+          if (lo(1) == domlo(1)) then
+
+             do k = lo(3)-dg(3), hi(3)+dg(3)
+                do j = lo(2)-dg(2), hi(2)+dg(2)
+
+                   ! reset the left state at domlo(1) if needed -- it is outside the domain
+                   if (physbc_lo(1) == Symmetry) then
```

Comment:
Perhaps we should do these if checks outside the loop?

Reply:
I fixed this